### ISSUE TITLE
Seems incompatible with Keycloak 21

### ISSUE THREAD

**Issue by JakeTheSnake** [aerogear/keycloak-metrics-spi#155]
## Description
After having migrated to Keycloak 21 (Quarkus) I get "Internal Server Error" whenever I try to login. The logs show messages like these:

`ERROR [io.quarkus.vertx.http.runtime.QuarkusErrorHandler] (executor-thread-7) HTTP Request to /auth/realms/jake/metrics failed, error id: fdebc248-5921-4c51-8db2-2bcbcff3607e-57: org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: Failed to register Collector of type Counter: keycloak_login_attempts_total is already in use by another Collector of type Counter`

Keycloak 21 changed metrics implementation from SmallRye to Micrometer, maybe this clashes with this project somehow?

## Steps to reproduce
1. Install and enable keycloak-metrics-spi v2.5.3 on Keycloak 20
2. Migrate to Keycloak 21
3. Attempt to login to the admin console

**Comment by logopk** [comment]
I noticed this too, but I thought it might only affect the metrics call from prometheus (thus I disabled the provider).

**Comment by logopk** [comment]
Caused by #92

**Comment by JaniszM** [comment]
Same problem here 👍

**Comment by Ekouyoja** [comment]
Confirm, same problem here

**Comment by fawadasaurus** [comment]
It appears that key cloak now natively incudes a /metrics endpoint in prometheus format with Keycloak 21 so it may be that this project is no longer needed for Keycloak 21+:
https://www.keycloak.org/2023/02/keycloak-2100-released.html

**Comment by logopk** [comment]
I don't see any metrics for specific realms in there - just the ones that existed before (strangely more or less consitently renamed ) .

**Comment by Jeroen-Van-Loocke** [comment]
I seem to have the same problem but already have issues when starting Keycloak.
This is what i get from the logs:

2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered global MeterFilters : [io.micrometer.core.instrument.config.MeterFilter$9@7573c7b5, io.micrometer.core.instrument.config.MeterFilter$9@73da7f08, io.micrometer.core.instrument.config.MeterFilter$9@717a8a76, io.micrometer.core.instrument.config.MeterFilter$9@62de73eb]
2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Configuring Micrometer registries : [class io.micrometer.prometheus.PrometheusMeterRegistry]
2023-03-02 10:31:22,776 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Discovered MeterFilters for class io.micrometer.prometheus.PrometheusMeterRegistry: []
2023-03-02 10:31:22,792 DEBUG [io.quarkus.micrometer.runtime.MicrometerRecorder] (main) Adding configured registry class io.micrometer.prometheus.PrometheusMeterRegistry
2023-03-02 10:31:23,385 ERROR [org.infinispan.CONFIG] (keycloak-cache-init) ISPN000660: DefaultCacheManager start failed, stopping any running components: org.infinispan.commons.CacheConfigurationException: ISPN000085: Error while trying to create a channel using the specified configuration file: default-configs/default-jgroups-udp.xml
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:758)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.initChannel(JGroupsTransport.java:504)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.start(JGroupsTransport.java:485)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:42)
	at org.infinispan.remoting.transport.jgroups.CorePackageImpl$1.start(CorePackageImpl.java:27)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.metrics.impl.MetricsCollector.start(MetricsCollector.java:78)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:41)
	at org.infinispan.metrics.impl.CorePackageImpl$1.start(CorePackageImpl.java:34)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.invokeStart(BasicComponentRegistryImpl.java:617)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:608)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startDependencies(BasicComponentRegistryImpl.java:635)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.doStartWrapper(BasicComponentRegistryImpl.java:599)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl.startWrapper(BasicComponentRegistryImpl.java:577)
	at org.infinispan.factories.impl.BasicComponentRegistryImpl$ComponentWrapper.running(BasicComponentRegistryImpl.java:808)
	at org.infinispan.factories.AbstractComponentRegistry.internalStart(AbstractComponentRegistry.java:357)
	at org.infinispan.factories.AbstractComponentRegistry.start(AbstractComponentRegistry.java:250)
	at org.infinispan.manager.DefaultCacheManager.internalStart(DefaultCacheManager.java:774)
	at org.infinispan.manager.DefaultCacheManager.start(DefaultCacheManager.java:742)
	at org.infinispan.manager.DefaultCacheManager.<init>(DefaultCacheManager.java:406)
	at org.keycloak.quarkus.runtime.storage.legacy.infinispan.CacheManagerFactory.startCacheManager(CacheManagerFactory.java:96)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.IllegalArgumentException: failed initializing component ThreadPool in protocol UDP: java.lang.IllegalStateException: failed to create virtual thread pool: java.lang.UnsupportedOperationException: Preview Features not enabled, need to run with --enable-preview
	at org.jgroups.stack.ProtocolStack.lambda$initComponents$7(ProtocolStack.java:861)
	at org.jgroups.util.Util.forAllComponents(Util.java:3368)
	at org.jgroups.stack.ProtocolStack.initComponents(ProtocolStack.java:839)
	at org.jgroups.stack.ProtocolStack.initProtocolStack(ProtocolStack.java:795)
	at org.jgroups.stack.ProtocolStack.setup(ProtocolStack.java:438)
	at org.jgroups.JChannel.init(JChannel.java:899)
	at org.jgroups.JChannel.<init>(JChannel.java:123)
	at org.jgroups.JChannel.<init>(JChannel.java:113)
	at org.infinispan.remoting.transport.jgroups.JGroupsTransport.buildChannel(JGroupsTransport.java:756)
	... 29 more

Hopefully this helps in finding what is wrong.

**Comment by gbatalski** [comment]
There are no keycloak specific metrics within the default implementation, so these metrics were different

**Comment by sschnabe** [comment]
IHMO this repository is abandoned.

We relied on this implementation, but decided to rewrite it for Keycloak 21: [kokuwaio/keycloak-event-metrics](https://github.com/kokuwaio/keycloak-event-metrics) This implementation adds event metrics to Micrometer from Quarkus at `/metrics`.

Differences:
 - no Prometheus push (event listener only adds counter to Micrometer)
 - no realm specific Prometheus endpoint, only `/metrics`
 - no http metrics, this is [already](https://www.keycloak.org/server/configuration-metrics#_available_metrics) included in Keycloak
 - different metric names, see [readme](https://github.com/kokuwaio/keycloak-event-metrics#readme)

Feel free to provide feedback.

**Comment by pb82** [comment]
thanks @Bilbergia I wonder if there is a way to make this work on both Keycloak 21 and earlier versions. We could catch the exception and carry on, but could it happen that the SPI initializes first and Keycloak fails registering its exporter?

**Comment by ghenadiibatalski** [comment]
@logopk i only needed to remove `_total` from the metric name. as i understand, `_total` is a reserved suffix now. The keycloak 21 ships the prometheus 0.16.0 so i updated the PR to keycloak 21 I don't know, if it could be backported for both versions. But the previous build is good enough for older Keycloak and the new one is compatible from 21 upwards.

**Comment by pb82** [comment]
@ghenadiibatalski has submitted a fix and I've verified it with KC 19, 20 and 21. I want to get it merged and released as soon as possible.

**Comment by logopk** [comment]
@pb82 shouldn't pom.xml be updated to to go back to Java 11?

**Comment by logopk** [comment]
Thanks @pb82

**Comment by souravs17031999** [comment]
@pb82 @JakeTheSnake I am still getting same errors as originally reported even after upgrading to 3.0.0 with KC 21.1.1

```
2023-04-29 23:15:47,442 ERROR [io.quarkus.vertx.http.runtime.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /favicon.ico failed, error id: 04833654-12f0-4d3b-a0a7-c910a8ba064e-4: org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: Failed to register Collector of type Counter: keycloak_login_attempts_total is already in use by another Collector of type Counter
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:230)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:258)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:430)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:408)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:140)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$0(QuarkusRequestFilter.java:82)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalArgumentException: Failed to register Collector of type Counter: keycloak_login_attempts_total is already in use by another Collector of type Counter
        at io.prometheus.client.CollectorRegistry.register(CollectorRegistry.java:57)
        at io.prometheus.client.SimpleCollector$Builder.register(SimpleCollector.java:261)
        at io.prometheus.client.SimpleCollector$Builder.register(SimpleCollector.java:253)
        at org.jboss.aerogear.keycloak.metrics.PrometheusExporter.<init>(PrometheusExporter.java:80)
        at org.jboss.aerogear.keycloak.metrics.PrometheusExporter.instance(PrometheusExporter.java:214)
        at org.jboss.aerogear.keycloak.metrics.MetricsFilter.filter(MetricsFilter.java:58)
        at org.jboss.aerogear.keycloak.metrics.MetricsFilterProvider.filter(MetricsFilterProvider.java:22)
        at org.jboss.resteasy.core.interception.jaxrs.ContainerResponseContextImpl.filter(ContainerResponseContextImpl.java:367)
        at org.jboss.resteasy.core.ServerResponseWriter.executeFilters(ServerResponseWriter.java:252)
        at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:101)
        at org.jboss.resteasy.core.ServerResponseWriter.writeNomapResponse(ServerResponseWriter.java:74)
        at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:226)
        ... 30 more
```

Any ideas how can I unblock myself ?

**Comment by amir4895** [comment]
@pb82 @ghenadiibatalski did this issue was fixed in a release for KC 21? i tried 21.1.2 but i still see the counter issue
 when trying to creat new realm 


org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: Failed to register Collector of type Counter: keycloak_login_attempts_total is already in use by another Collector of type Counter
